content/blog/taming-react-re-renders.md

```diff
@@ -0,0 +1,174 @@
++++
+title = "Taming React Re-renders: A Guide to Optimizing Performance"
+slug = "taming-react-re-renders"
+date = 2025-06-11T20:48:34+05:30
+image = "/images/2025/taming-react-re-renders/header.jpg"
+draft = true
+authors = ["Ajith Kumar"]
+description = "A comprehensive guide to understanding and optimizing React re-renders for better application performance"
+tags = ["React", "Performance", "Software Craftsmanship"]
+categories = ["React", "Performance", "Software Craftsmanship"]
+type = ""
++++
+
+Have you ever noticed your React app feeling sluggish? or wondered why your components keep re-rendering when they shouldn't? You're not alone!
+
+## The Problem: Unnecessary Re-renders
+
+Let's start with a simple example. Imagine you're building a todo list app:
+
+```jsx
+function TodoApp() {
+  const [todos, setTodos] = useState([]);
+  const [text, setText] = useState('');
+
+  const addTodo = () => {
+    setTodos([...todos, text]);
+    setText('');
+  };
+
+  return (
+    <div>
+      <input
+        value={text}
+        onChange={(e) => setText(e.target.value)}
+        placeholder='Add todo…'
+      />
+      <button onClick={addTodo}>Add</button>
+      <TodoList items={todos} />
+    </div>
+  );
+}
+
+function TodoList({items}) {
+  console.log('Rendering list');
+  return (
+    <ul>
+      {items.map((item) => (
+        <li key={item.id}>{item.text}</li>
+      ))}
+    </ul>
+  );
+}
+```
+
+Notice something interesting? Every time you type in the input field, the entire `TodoList` re-renders, even though you haven't added any new todos! This happens because React re-renders the parent component (`TodoApp`) when its state changes, which then re-renders all its children.
+
+## Understanding React's Lifecycle
+
+To fix unnecessary re-renders, we need to understand how React works. Think of React components like a tree:
+
+- When a component first appears on screen, it **mounts**
+- When its data (state or props) changes, it **updates**
+- When it's removed from the screen, it **unmounts**
+
+Here's what you need to know:
+
+- **State changes** trigger re-renders
+
+{{< figure src="/images/2025/taming-react-re-renders/state-changes-example.jpg" caption="" >}}
```

Comment:
suggestion: will be good to add some explanation after each of the images, giving for details into what the image is talking about